pandas/core/internals/managers.py

```diff
@@ -59,7 +62,867 @@
+                    arr = self._make_na_array(fill_value=fill_value)
+                else:
+                    arr = self.arrays[i]
+                new_arrays.append(arr)
+
+        else:
+            new_arrays = [
+                algos.take(
+                    arr,
+                    indexer,
+                    allow_fill=True,
+                    fill_value=fill_value,
+                    # if fill_value is not None else blk.fill_value
+                )
+                for arr in self.arrays
+            ]
+
+        new_axes = list(self._axes)
+        new_axes[axis] = new_axis
+
+        return type(self)(new_arrays, new_axes)
+
+    def take(self, indexer, axis: int = 1, verify: bool = True, convert: bool = True):
+        """
+        Take items along any axis.
+        """
+        axis = self._normalize_axis(axis)
+
+        indexer = (
+            np.arange(indexer.start, indexer.stop, indexer.step, dtype="int64")
+            if isinstance(indexer, slice)
+            else np.asanyarray(indexer, dtype="int64")
+        )
+
+        n = self.shape_proper[axis]
+        if convert:
+            indexer = maybe_convert_indices(indexer, n)
+
+        if verify:
+            if ((indexer == -1) | (indexer >= n)).any():
+                raise Exception("Indices must be nonzero and less than the axis length")
+
+        new_labels = self._axes[axis].take(indexer)
+        return self._reindex_indexer(
+            new_axis=new_labels, indexer=indexer, axis=axis, allow_dups=True
+        )
+
+    def _make_na_array(self, fill_value=None):
+        if fill_value is None:
+            fill_value = np.nan
+
+        dtype, fill_value = infer_dtype_from_scalar(fill_value)
+        values = np.empty(self.shape_proper[0], dtype=dtype)
+        values.fill(fill_value)
+        return values
+
+    def unstack(self, unstacker, fill_value) -> "ArrayManager":
+        """
+        Return a BlockManager with all blocks unstacked..
+
+        Parameters
+        ----------
+        unstacker : reshape._Unstacker
+        fill_value : Any
+            fill_value for newly introduced missing values.
+
+        Returns
+        -------
+        unstacked : BlockManager
+        """
+        indexer, _ = unstacker._indexer_and_to_sort
+        new_indexer = np.full(unstacker.mask.shape, -1)
+        new_indexer[unstacker.mask] = indexer
+        new_indexer2D = new_indexer.reshape(*unstacker.full_shape)
+
+        new_arrays = []
+        for arr in self.arrays:
+            for i in range(unstacker.full_shape[1]):
+                new_arr = algos.take(
+                    arr, new_indexer2D[:, i], allow_fill=True, fill_value=fill_value
+                )
+                new_arrays.append(new_arr)
+
+        new_index = unstacker.new_index
+        new_columns = unstacker.get_new_columns(self._axes[1])
+        new_axes = [new_index, new_columns]
+
+        return type(self)(new_arrays, new_axes, do_integrity_check=False)
+
```

Comment:
can you put all of this in a new file: array_manager (ok to rename the existing to block_manager)
OR add another level of nesting (panda/core/internals/managers/array, block) either way ok.